File: 0010.RegularExpressionMatching/RegularExpressionMatching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        slen, plen = len(s), len(p)
        if slen == 0 or plen == 0:
            return self.isMatch(s, p[2:]) if plen >= 2 and p[1] == '*' else slen == plen
        if plen == 1 or p[1] != '*':
            return False if p[0] != '.' and s[0] != p[0] else self.isMatch(s[1:], p[1:])
        return self.isMatch(s[1:], p) or self.isMatch(s, p[2:]) if s[0] == p[0] or p[0] == '.' else self.isMatch(s, p[2:])

File: 0010.RegularExpressionMatching/RegularExpressionMatching.c
int isMatch(char *s, char *p)
{
        while (*p && *s)
        {
                if (p[1] != '*')
                {
                        if (*p == '.' || *s == *p)
                                p++, s++;
                        else
                                return 0;
                }
                else
                {
                        if (*p != '.' && *s != *p)
                                p += 2;
                        else
                                return isMatch(s + 1, p) || isMatch(s, p + 2);
                }
        }

        while (*p && p[1] == '*')
                p += 2;
        return *s == *p;
}